### PR TITLE
Add back the attacking entity to LivingKnockBackEvent 

### PR DESCRIPTION
**Pull request by Sarinsa**
Add back the attacking entity to LivingKnockBackEvent and add nulllable entity parameter for attacking entity to LivingEntity#knockback

### CODE REVIEW

File: patches/minecraft/net/minecraft/entity/LivingEntity.java.patch [hunk 1]
```diff
@@ -213,11 +231,13 @@
...
  
     protected LootContext.Builder func_213363_a(boolean p_213363_1_, DamageSource p_213363_2_) {
-@@ -1273,6 +_,11 @@
+@@ -1272,7 +_,12 @@
+       return lootcontext$builder;
     }
  
-    public void func_233627_a_(float p_233627_1_, double p_233627_2_, double p_233627_4_) {
-+      net.minecraftforge.event.entity.living.LivingKnockBackEvent event = net.minecraftforge.common.ForgeHooks.onLivingKnockBack(this, p_233627_1_, p_233627_2_, p_233627_4_);
+-   public void func_233627_a_(float p_233627_1_, double p_233627_2_, double p_233627_4_) {
```
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/entity/LivingEntity.java.patch:238]: We do not remove methods, we deprecate them and defer the call.

File: patches/minecraft/net/minecraft/entity/LivingEntity.java.patch [hunk 2]
```diff
@@ -213,11 +231,13 @@
...
++   public void knockback(@Nullable Entity attacker, float p_233627_1_, double p_233627_2_, double p_233627_4_) {
++      net.minecraftforge.event.entity.living.LivingKnockBackEvent event = net.minecraftforge.common.ForgeHooks.onLivingKnockBack(attacker,this, p_233627_1_, p_233627_2_, p_233627_4_);
```
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/entity/LivingEntity.java.patch:240]: Space between `attacker` and `this`.

File: patches/minecraft/net/minecraft/entity/LivingEntity.java.patch [hunk 3]
```diff
@@ -147,6 +147,24 @@
...
++               this.knockback(p_70097_1_.func_76364_f(), 0.4F, d1, d0);
+             } else {
+                this.field_70739_aP = (float)((int)(Math.random() * 2.0D) * 180);
+             }
+@@ -1073,7 +_,7 @@
+    }
+ 
+    protected void func_213371_e(LivingEntity p_213371_1_) {
+-      p_213371_1_.func_233627_a_(0.5F, p_213371_1_.func_226277_ct_() - this.func_226277_ct_(), p_213371_1_.func_226281_cx_() - this.func_226281_cx_());
++      p_213371_1_.knockback(this,0.5F, p_213371_1_.func_226277_ct_() - this.func_226277_ct_(), p_213371_1_.func_226281_cx_() - this.func_226281_cx_());
```
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/entity/LivingEntity.java.patch:164]: Space between `this` and `0.5F`

File: patches/minecraft/net/minecraft/entity/MobEntity.java.patch [hunk 4]
```diff
@@ -50,3 +50,12 @@
           }
        } else {
           return EquipmentSlotType.HEAD;
+@@ -1353,7 +_,7 @@
+       boolean flag = p_70652_1_.func_70097_a(DamageSource.func_76358_a(this), f);
+       if (flag) {
+          if (f1 > 0.0F && p_70652_1_ instanceof LivingEntity) {
+-            ((LivingEntity)p_70652_1_).func_233627_a_(f1 * 0.5F, (double)MathHelper.func_76126_a(this.field_70177_z * ((float)Math.PI / 180F)), (double)(-MathHelper.func_76134_b(this.field_70177_z * ((float)Math.PI / 180F))));
++            ((LivingEntity)p_70652_1_).knockback(this,f1 * 0.5F, (double)MathHelper.func_76126_a(this.field_70177_z * ((float)Math.PI / 180F)), (double)(-MathHelper.func_76134_b(this.field_70177_z * ((float)Math.PI / 180F))));
```
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/entity/MobEntity.java.patch:58]: Space between `this` and `f1 * 0.5F`

File: patches/minecraft/net/minecraft/entity/player/PlayerEntity.java.patch [hunk 5]
```diff
@@ -229,6 +229,24 @@
                 }
  
                 f = f + f1;
+@@ -1068,7 +_,7 @@
+                if (flag5) {
+                   if (i > 0) {
+                      if (p_71059_1_ instanceof LivingEntity) {
+-                        ((LivingEntity)p_71059_1_).func_233627_a_((float)i * 0.5F, (double)MathHelper.func_76126_a(this.field_70177_z * ((float)Math.PI / 180F)), (double)(-MathHelper.func_76134_b(this.field_70177_z * ((float)Math.PI / 180F))));
++                        ((LivingEntity)p_71059_1_).knockback(this,(float)i * 0.5F, (double)MathHelper.func_76126_a(this.field_70177_z * ((float)Math.PI / 180F)), (double)(-MathHelper.func_76134_b(this.field_70177_z * ((float)Math.PI / 180F))));
```
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/entity/player/PlayerEntity.java.patch:237]: Space after `this` comma.

File: patches/minecraft/net/minecraft/entity/player/PlayerEntity.java.patch [hunk 6]
```diff
@@ -229,6 +229,24 @@
...
+                         p_71059_1_.func_70024_g((double)(-MathHelper.func_76126_a(this.field_70177_z * ((float)Math.PI / 180F)) * (float)i * 0.5F), 0.1D, (double)(MathHelper.func_76134_b(this.field_70177_z * ((float)Math.PI / 180F)) * (float)i * 0.5F));
+                      }
+@@ -1082,7 +_,7 @@
+ 
+                      for(LivingEntity livingentity : this.field_70170_p.func_217357_a(LivingEntity.class, p_71059_1_.func_174813_aQ().func_72314_b(1.0D, 0.25D, 1.0D))) {
+                         if (livingentity != this && livingentity != p_71059_1_ && !this.func_184191_r(livingentity) && (!(livingentity instanceof ArmorStandEntity) || !((ArmorStandEntity)livingentity).func_181026_s()) && this.func_70068_e(livingentity) < 9.0D) {
+-                           livingentity.func_233627_a_(0.4F, (double)MathHelper.func_76126_a(this.field_70177_z * ((float)Math.PI / 180F)), (double)(-MathHelper.func_76134_b(this.field_70177_z * ((float)Math.PI / 180F))));
++                           livingentity.knockback(this,0.4F, (double)MathHelper.func_76126_a(this.field_70177_z * ((float)Math.PI / 180F)), (double)(-MathHelper.func_76134_b(this.field_70177_z * ((float)Math.PI / 180F))));
```
Comment (ChampionAsh5357) [inline patches/minecraft/net/minecraft/entity/player/PlayerEntity.java.patch:246]: Space after `this` comma.

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 7]
```diff
@@ -350,9 +350,9 @@ public static boolean onPlayerAttack(LivingEntity entity, DamageSource src, floa
         return !MinecraftForge.EVENT_BUS.post(new LivingAttackEvent(entity, src, amount));
     }
 
-    public static LivingKnockBackEvent onLivingKnockBack(LivingEntity target, float strength, double ratioX, double ratioZ)
+    public static LivingKnockBackEvent onLivingKnockBack(@Nullable Entity attacker, LivingEntity target, float strength, double ratioX, double ratioZ)
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java:353]: Breaking change, should have another method which is deprecated which delegates to this one.

File: src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java [hunk 8]
```diff
@@ -26,15 +26,14 @@
...
 
+import javax.annotation.Nullable;
+
 /**
  * LivingKnockBackEvent is fired when a living entity is about to be knocked back. <br>
  * This event is fired whenever an Entity is knocked back in
- * {@link EntityLivingBase#attackEntityFrom(DamageSource, float)},
- * {@link EntityLivingBase#blockWithShield(EntityLivingBase)},
- * {@link EntityMob#attackEntityAsMob(Entity)} and
- * {@link EntityPlayer#attackTargetEntityWithCurrentItem(Entity)} <br>
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java:35]: Why? They are still being called at these places.

File: src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java [hunk 9]
```diff
@@ -51,19 +50,26 @@
...
+    @Nullable
+    protected Entity attacker;
+
     protected float strength;
     protected double ratioX, ratioZ;
     protected final float originalStrength;
     protected final double originalRatioX, originalRatioZ;
 
-    public LivingKnockBackEvent(LivingEntity target, float strength, double ratioX, double ratioZ)
+    public LivingKnockBackEvent(@Nullable Entity attacker, LivingEntity target, float strength, double ratioX, double ratioZ)
```
Comment (ChampionAsh5357) [inline src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java:61]: Breaking change, deprecate old constructor for removal and create new one. Old constructor delegates to new one.

**Review by ChampionAsh5357 — CHANGES_REQUESTED** [review]
Since 1.16 is now LTS, PRs should be made to 1.17 first if applicable and then backported when merged in.

File: src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java (hunk 9 above)
Comment (Sarinsa) [reply src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java:61]: Yeah, I see I have been way too hasty with this

**Comment by sciwhiz12** [comment]
Hello! @Sarinsa, please leave a comment whether you wish to continue this PR and update it to the latest active version of 1.18 (or leave an explanation why this PR's changes are not applicable to 1.18). If we do not hear back from you after some time, this PR will be closed for inactivity.
If anyone else wishes to continue the work of this PR, please create a PR with a link to this PR for reference. 
Thank you for your contributions. 👋

**Comment by sciwhiz12** [comment]
Hello! Since the application of the Stale label, there has been no recent activity on this PR. As said in our previous comment, this PR will now be closed for inactivity.

If you wish to continue this PR, please open a new PR against the Latest supported version (to be conformant with our version support and contributions policy). 

Thank you for your contributions. 👋

**Pull request closed without merging**